### ISSUE TITLE
node-pre-gyp install error

### ISSUE THREAD

**Issue by zjuwjf** [mapbox/node-pre-gyp#413]
zjuwjfdeMacBook-Pro:dsl2 zjuwjf$ tnpm install canvas@next
✔ Installed 1 packages
✔ Linked 66 latest versions
node-pre-gyp http GET https://github.com/node-gfx/node-canvas-prebuilt/releases/download/v2.0.0-alpha.13/canvas-prebuilt-v2.0.0-alpha.13-node-v64-darwin-unknown-x64.tar.gz
node-pre-gyp http 302 https://github.com/node-gfx/node-canvas-prebuilt/releases/download/v2.0.0-alpha.13/canvas-prebuilt-v2.0.0-alpha.13-node-v64-darwin-unknown-x64.tar.gz
node-pre-gyp ERR! Completion callback never invoked!
node-pre-gyp ERR! System Darwin 16.7.0
node-pre-gyp ERR! command "/usr/local/Cellar/node/10.1.0/bin/node" "/Users/zjuwjf/idlefish-git/dsl2/node_modules/_canvas@2.0.0-alpha.13@canvas/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /Users/zjuwjf/idlefish-git/dsl2/node_modules/_canvas@2.0.0-alpha.13@canvas
node-pre-gyp ERR! node -v v10.3.0
node-pre-gyp ERR! node-pre-gyp -v v0.9.1
node-pre-gyp ERR! This is a bug in `node-pre-gyp`.
node-pre-gyp ERR! Try to update node-pre-gyp and file an issue if it does not help:
node-pre-gyp ERR!     <https://github.com/mapbox/node-pre-gyp/issues>
✖ Install fail! Error: post install error, please remove node_modules before retry!
Run "sh -c node-pre-gyp install --fallback-to-build" error, exit code 6
Error: post install error, please remove node_modules before retry!
Run "sh -c node-pre-gyp install --fallback-to-build" error, exit code 6
    at ChildProcess.proc.on.code (/usr/local/lib/node_modules/tnpm/node_modules/_runscript@1.3.0@runscript/index.js:74:21)
    at ChildProcess.emit (events.js:182:13)
    at maybeClose (internal/child_process.js:961:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:248:5)
npminstall version: 3.7.0
npminstall args: /usr/local/Cellar/node/10.1.0/bin/node /usr/local/lib/node_modules/tnpm/node_modules/_npminstall@3.7.0@npminstall/bin/install.js --tmp=/Users/zjuwjf/.tnpm_tmp --no-proxy --disturl=https://tnpm-hz.oss-cn-hangzhou.aliyuncs.com/dist/node --userconfig=/Users/zjuwjf/.tnpmrc --registry=http://registry.npm.alibaba-inc.com canvas@next --china --prefix=/usr/local --forbidden-licenses=AGPL,CPAL,OSL --fix-bug-versions
[tnpm] subprocess:/usr/local/lib/node_modules/tnpm/node_modules/_npminstall@3.7.0@npminstall/bin/install.js exit code: 1, signal: null
[tnpm] mainprocess exit code: 1

node：v10.3.0

**Comment by ilatypov** [comment]
I wonder if attempts to connect to GitHub's Amazon storage failed due to a firewall.  (Also, I could not find the source code of tnpm).

**Comment by smadds** [comment]
I'm getting the same error - any ideas anyone (it's not a firewall issue on my side)

**Comment by jmhiggins42** [comment]
Try `tnpm install canvas@next --build-from-source`. That should pass the --build-from-source option to node-pre-gyp, which skips trying to get the binaries all together.

The command is failing here:

>node-pre-gyp http GET https://github.com/node-gfx/node-canvas-prebuilt/releases/download/v2.0.0-alpha.13/canvas-prebuilt-v2.0.0-alpha.13-node-v64-darwin-unknown-x64.tar.gz
>node-pre-gyp http 302 https://github.com/node-gfx/node-canvas-prebuilt/releases/download/v2.0.0-alpha.13/canvas-prebuilt-v2.0.0-alpha.13-node-v64-darwin-unknown-x64.tar.gz

Not sure about you, but I've been finding that my machine has some TLS-related issues. For example, I can't `pip install` anything (I get 403 Forbidden errors), I must use `sudo easy_install`. So being able to bypass this step in the install process did the trick.

**Comment by jmhiggins42** [comment]
No problem - I put a decent amount of research in this weekend on how all this node-gyp / node-pre-gyp business works and I'm finally getting the hang of it.

@zjuwjf had mentioned `tnpm`, hence why I responded with that. I imagine it works a lot like npm, but with whatever peculiarities it has.

As for your problem now, these are the lines to pay attention to:

>/bin/sh: 1: pkg-config: not found
>gyp: Call to 'pkg-config pixman-1 --libs' returned exit status 127 while in binding.gyp. while trying to load binding.gyp

These are lines 126-131 of this canvas@next's [binding.gyp](https://github.com/Automattic/node-canvas/blob/master/binding.gyp) file:

 ```python
   'libraries': [
            '<!@(pkg-config pixman-1 --libs)',
            '<!@(pkg-config cairo --libs)',
            '<!@(pkg-config libpng --libs)',
            '<!@(pkg-config pangocairo --libs)',
            '<!@(pkg-config freetype2 --libs)'
          ],
```

This is trying to link the necessary libraries and assumes you have each of them (pixman-1, cairo, libpng etc.) already installed. For one reason or another, the CLI command `pkg-config pixman-1 --libs` is failing. You may need to install the package before you can install canvas, or you might just need to tell pkg-config where to find the pixman-1 library. When I tried using `pkg-config xxx --libs` in my Ubuntu shell, this is what I got back:

>Perhaps you should add the directory containing `xxx.pc'
>to the PKG_CONFIG_PATH environment variable

So you might just need to add the directories of these pc files to PKG_CONFIG_PATH in you .bashrc/.zshrc file:

```shell
export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:path/to/lib/A:path/to/lib/B
```

**TL;DR** First follow the [Compiling instructions](https://github.com/Automattic/node-canvas#compiling) for canvas@next. If that doesn't fix it, run `pkg-config xxx --libs` in your command line to see if they succeed for the packages listed above. If those fail, try the commands `cd /` and `find . -name xxx.pc`. That takes you to your root directory and scans the entire drive for the particular .pc file. Once you've found the directory, update your PKG_CONFIG_PATH env variable. And if all that fails, then 🤷‍♂️. But respond back with the error message and I can try to help.

**Comment by smadds** [comment]
The problem was the RPi didn't have pkg-config installed - once I got that everything worked - with the odd warning...
Thanks for your support, John.

```
node-pre-gyp install --fallback-to-build

node-pre-gyp WARN Using needle for node-pre-gyp https download 
node-pre-gyp WARN Tried to download(404): https://github.com/tessel/node-usb/releases/download/1.3.3/usb_bindings-v1.3.3-node-v57-linux-arm.tar.gz 
node-pre-gyp WARN Pre-built binaries not found for usb@1.3.3 and node@8.12.0 (node-v57 ABI, glibc) (falling back to source compile with node-gyp) 
make: Entering directory '/home/simon.maddox/.node-red/node_modules/usb/build'
  CC(target) Release/obj.target/libusb/libusb/libusb/core.o
  CC(target) Release/obj.target/libusb/libusb/libusb/descriptor.o
  CC(target) Release/obj.target/libusb/libusb/libusb/hotplug.o
  CC(target) Release/obj.target/libusb/libusb/libusb/io.o
  CC(target) Release/obj.target/libusb/libusb/libusb/strerror.o
  CC(target) Release/obj.target/libusb/libusb/libusb/sync.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/poll_posix.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/threads_posix.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/linux_usbfs.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/linux_udev.o
  AR(target) Release/obj.target/usb.a
  COPY Release/usb.a
  CXX(target) Release/obj.target/usb_bindings/src/node_usb.o
../src/node_usb.cc: In function ‘void handleHotplug(std::pair<libusb_device*, libusb_hotplug_event>)’:
../src/node_usb.cc:151:58: warning: ‘v8::Local<v8::Value> Nan::MakeCallback(v8::Local<v8::Object>, const char*, int, v8::Local<v8::Value>*)’ is deprecated [-Wdeprecated-declarations]
  Nan::MakeCallback(Nan::New(hotplugThis), "emit", 2, argv);
                                                          ^
In file included from ../src/helpers.h:3:0,
                 from ../src/node_usb.h:21,
                 from ../src/node_usb.cc:1:
../../nan/nan.h:1000:46: note: declared here
   NAN_DEPRECATED inline v8::Local<v8::Value> MakeCallback(
                                              ^~~~~~~~~~~~
  CXX(target) Release/obj.target/usb_bindings/src/device.o
../src/device.cc: In static member function ‘static void Req::default_after(uv_work_t*)’:
../src/device.cc:237:64: warning: ‘v8::Local<v8::Value> Nan::MakeCallback(v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*)’ is deprecated [-Wdeprecated-declarations]
    Nan::MakeCallback(device, Nan::New(baton->callback), 1, argv);
                                                                ^
In file included from ../src/helpers.h:3:0,
                 from ../src/node_usb.h:21,
                 from ../src/device.cc:1:
../../nan/nan.h:958:46: note: declared here
   NAN_DEPRECATED inline v8::Local<v8::Value> MakeCallback(
                                              ^~~~~~~~~~~~
  CXX(target) Release/obj.target/usb_bindings/src/transfer.o
../src/transfer.cc: In function ‘void handleCompletion(Transfer*)’:
../src/transfer.cc:126:72: warning: ‘v8::Local<v8::Value> Nan::MakeCallback(v8::Local<v8::Object>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*)’ is deprecated [-Wdeprecated-declarations]
   Nan::MakeCallback(self->handle(), Nan::New(self->v8callback), 3, argv);
                                                                        ^
In file included from ../src/helpers.h:3:0,
                 from ../src/node_usb.h:21,
                 from ../src/transfer.cc:1:
../../nan/nan.h:958:46: note: declared here
   NAN_DEPRECATED inline v8::Local<v8::Value> MakeCallback(
                                              ^~~~~~~~~~~~
  SOLINK_MODULE(target) Release/obj.target/usb_bindings.node
  COPY Release/usb_bindings.node
  COPY /home/simon.maddox/.node-red/node_modules/usb/src/binding/usb_bindings.node
  TOUCH Release/obj.target/action_after_build.stamp
make: Leaving directory '/home/simon.maddox/.node-red/node_modules/usb/build'

```

**Comment by fifahuihua** [comment]
try to remove or rename the folder of openssl in your current user's home folder. 
For example: C:\Users\zhangan2\.node-gyp\9.11.1\include\node

**Comment by hunnimonstr** [comment]
hi i have been trying to restore my node-red system on my raspberry Pi-0-W for a few days now with successive failures due to unavailable resources for node-pre-gyp
I started with the current version of node.js and node red installed via the nodered.org pi page.
Hit the problem looked it up and found some folk using a Pi3b or 3B+ had success after downgrading to node8. so i started with a new raspbian-lite install and installed node.js 8 and then node red, installed my nodes and again no node-pre-gyp resourses resulting in a rolback/fallback build with known issues in the nobel/debug resources that the node i need 'node-red-contrib-Xiaomi-ble' fails to install with resulting in an unresolvable npm audit err..

I have now followed Pete scargills suggestion to downgrade to node 6, --  still the same issue..

 an attempt to install pkg-config failed too raising a screen full of Blood red compile errors before crashing out with exit condition 1.

On a fresh install the errors are reported around problems with bluetooth-hci-socket,  as i start trying to fix this with attempts to install bluetooth-hci-socket and usb and their respective prerequisites the errors soon solidify to 

-pre-gyp https download
node-pre-gyp WARN Tried to download(404): https://github.com/tessel/node-usb/releases/download/1.5.0/usb_bindings-v1.5.0-node-v64-linux-arm.tar.gz
node-pre-gyp WARN Pre-built binaries not found for usb@1.5.0 and node@10.15.1 (node-v64 ABI, glibc) (falling back to source compile with node-gyp)

which will result in the nobel/debug V2.0 installed while the xiaomi-ble node requires the bug fixed V2.6.9 of nobel/debug to install..  

and around and around i have been going..  Please advise???

**Comment by janeksz** [comment]
Failed to execute '/usr/bin/node /usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/root/.node-red/node_modules/usb/src/binding/usb_bindings.node --module_name=usb_bindings --module_path=/root/.node-red/node_modules/usb/src/binding --napi_version=3 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v64' (1)


root@RPi3AP:~# sqlite3 -version
3.16.2 2017-01-06 16:32:41 a65a62893ca8319e89e48b8a38cf8a59c69a8209
root@RPi3AP:~# node -v
v10.15.1
root@RPi3AP:~# npm -v
6.7.0
root@RPi3AP:~#

**Comment by brainmobilizer** [comment]
You can try `npm install request` first and `npm install` again. 
Or you can add `request` package to devDependencies and `npm install`
